Add (union) discriminator support

Spea opened the pull request:
With these changes it is now possible to use (union) discriminators. A few examples which are supported might look like this 👇 

<details>
<summary>Examples</summary>


### Discriminators

```php
#[JMS\Discriminator(field: 'type', map: ['first' => FirstClass::class, 'second' => SecondClass::class])]
abstract class Discriminator
{
}

class FirstClass extends Discriminator
{
    public string $firstProperty;
}

class SecondClass extends Discriminator
{
    public int $secondProperty;
}
```

### Union Discriminators

```php
use JMS\Serializer\Annotation\UnionDiscriminator;

class JmsUnionTyping
{
    #[UnionDiscriminator(field: 'objectType', map: ['comment' => FirstClass::class, 'author' => SecondClass::class])]
    public FirstClass|SecondClass $property;
}

// ...

class PrimitiveUnionTyping
{
    public int|string|false|float|array|null $property;
}
```


</details>

To support this, it is necessary to use the latest changes from the `2.x` branch from the `metadata-parser` (including the changes necessary for union discriminators via https://github.com/liip/metadata-parser/pull/54).
So ideally this change should go into a new `3.x` branch where we then release a new major version, as doing a minor release will potentially break peoples implementation since a few things changed in the parse library.

Side note: There is a temporary commit in here which uses the changes from the other PR in the `metadata-parser` repo. It will be removed/adjusted, once merged and a new version exists.

Spea committed:
Use `PropertyTypeIterable` instead of `PropertyTypeArray`

The `PropertyTypeArray` class was removed in favor the `PropertyTypeIterable` one.

Spea committed:
Add (JMS) discriminator support

With this change, the `Discriminator` attribute from JMS is now fully supported.

Spea committed:
Add support for union (discriminator attribute)

With this change, it is now possible to use primitive union types (like `public int|bool $property`) or the more complex `#[UnionDiscriminator]` attribute from JMS.

Spea committed:
Update php-cs-fixer and apply fixes

Spea committed:
Adjust README to reflect renamed file

dbu commented:
very cool, thanks!

i will do php 8.5 and symfony 8 support on the 2.x branch, then add a 3.x branch where we should merge your changes.

before we tag the 3.0 release, we should look at the [deprecation](https://github.com/liip/serializer/blob/6284797e1c39f4d249060a9bf92b1d71c61b2328/src/Template/Deserialization.php#L221) and do the BC break to cleanup.
apart from that i am not aware of things we have to do, are there?

dbu commented on composer.json:
```diff
@@ -13,10 +13,17 @@
         "php": "^8.0",
         "ext-json": "*",
-        "liip/metadata-parser": "^1.2",
+        "liip/metadata-parser": "dev-suppport-discriminator",
```
```suggestion
        "liip/metadata-parser": "^2.0",
```

Spea replied:
I only saw a release for `1.2.1` but not for `2.0` on [the release page](https://github.com/liip/metadata-parser/releases), did something went wrong while creating the new `2.0` release?

dbu replied:
sorry, i meant to ask you if you think metadata-parser is ready to tag or if something is still pending.

can you rebase and solve conflicts here? using `^2.0@dev` as dependency? then we see if all is green, and i tag metadata-parser if nothing else is pending.

Spea replied:
Ah yeah, makes sense. Let me do that.

Spea replied:
Done that now. 
I want to work on enum support next, so for me it would be fine if we wait for that feature before tagging a new release, but we could also move it to a potential `2.1` version in the `metadata-parser`. Up to you :)

dbu replied:
lets release often :-) 
i tagged https://github.com/liip/metadata-parser/releases/tag/2.0.0

if you can add enum without a BC break we can do a 2.1.0 soon.

Spea replied:
Awesome, I adjusted the `composer.json` accordingly :)

dbu commented:
i tagged 2.6.2 and now created the 3.x branch. i changed the target, but github reports conflicts. i did apply the cs fixer in the 2.x version, so that is probably not needed anymore. and i added to the changelog - i added the 2.x and 3.x headings so that future changes in 2.x do not conflict with 3.x anymore.

can you please rebase this on the 3.x branch and solve the conflicts?

Spea committed:
Remove deprecated string type for `formats` argument

Passing a `string` for the `$formats` argument when rendering the assignment of a `DateTime` class was deprecated in `2.x` and thus got for `3.x`.

Spea commented:
> before we tag the 3.0 release, we should look at the [deprecation](https://github.com/liip/serializer/blob/6284797e1c39f4d249060a9bf92b1d71c61b2328/src/Template/Deserialization.php#L221) and do the BC break to cleanup. apart from that i am not aware of things we have to do, are there?

I removed the deprecation warning and adjusted the code around it. Other than that I think we are good here, couldn't find any other deprecations.

> i tagged 2.6.2 and now created the 3.x branch. i changed the target, but github reports conflicts. i did apply the cs fixer in the 2.x version, so that is probably not needed anymore. and i added to the changelog - i added the 2.x and 3.x headings so that future changes in 2.x do not conflict with 3.x anymore.
> 
> can you please rebase this on the 3.x branch and solve the conflicts?

I resolved the conflicts now, thanks for taking care of the new branch and the cs fixes!

dbu commented on CHANGELOG.md:
```diff
@@ -4,6 +4,9 @@
 
 # 3.0.0 (unreleased)
 
+* Update to liip/metadata-parser 2.x
```
the BC break is in metadata-parser, but i think it still makes sense to mention here to switch the annotation => attribute parser change we did (i also expanded the changelog in metadata-parser repository to be more explicit about it). i would tell to switch from annotation to attribute and link to the changelog of the 2.x branch for further changes.

Spea replied:
Yeah its a good idea to mention this. I added some additional information via e2521f7fc24761ef04c2bdb1d150f1abfb04460f and also added one point about the change for the naming strategy.

Spea committed:
Mention some BC breaks for `metadata-parser`

Those are the more critical BC breaks, so its worth mentioning them here as well. For all further changes, the `CHANGELOG.md` from the `metadata-parser` is also referenced.

dbu commented:
thanks! i tweaked the changelog entry a bit and merged on the CLI. thanks a lot for the contribution!

i tagged https://github.com/liip/serializer/releases/tag/3.0.0 now
